# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [34]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import make_pipeline
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report


## 1. Loading the dataset

In [14]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('science_data_large.csv')
print(data.head(15))
print(data.info())


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [16]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [19]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)
score = model.score(X_test, y_test)
coefficients = model.coef_
intercept = model.intercept_
equation = f"Size nm^3 = {intercept:.5f} + {coefficients[0]:.5f} * Temperature + {coefficients[1]:.5f} * KCl"
print(f'Model Score (R^2): {score}')
print(f'Regression Equation: {equation}')

sample = pd.DataFrame([[500, 500]], columns=["Temperature °C", "Mols KCL"])
sample_prediction = model.predict(sample)
print(f"Prediction for sample input [500°C, 500 mols]: {sample_prediction[0]:.2f} nm^3")



Model Score (R^2): 0.8552472077276095
Regression Equation: Size nm^3 = -409391.47958 + 866.14641 * Temperature + 1032.69507 * KCl
Prediction for sample input [500°C, 500 mols]: 540029.26 nm^3


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$
\text{Size}_{nm^3} = b_0 + b_1 \cdot \text{Temperature (°C)} + b_2 \cdot \text{Mols KCL}
$$
Where:
- \( b_0 \): Intercept
- \( b_1 \), \( b_2 \): Learned coefficients

Report on score and explain meaning: The linear regression model explains approximately 85.5% of the variation in slime size based on the temperature and amount of KCl indicating a strong model fit.

## 4. Use Cross Validation

In [20]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

cv_scores = cross_val_score(model, X, y, cv=5)
print(f'Cross-validation scores: {cv_scores}')
print(f'Mean CV Score: {cv_scores.mean()}')


Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean CV Score: 0.8568167899144437


Write findings here:
### Findings from Cross-Validation

The cross-validation scores obtained from the model are as follows:  
\[ [0.83918826, 0.87051239, 0.85871066, 0.87202623, 0.84364641] \]

The mean cross-validation score is approximately **0.8568**. 

#### Significance of Findings:
1. **Good Performance**: The mean R² score of around **0.857** indicates that the model explains approximately 85.7% of the variance in the size of the slime based on temperature and KCL concentration. This suggests that the model is fairly effective at capturing the relationship between these variables and the target size.

2. **Consistency Across Splits**: The range of the individual cross-validation scores (from **0.839** to **0.872**) indicates a relatively high consistency across different data splits, implying that the model's performance is reliable and not overly sensitive to the specific data used in training.

3. **Model Robustness**: Cross-validation helps in assessing the robustness of the model. Since the scores are all significantly positive, it indicates that the model is likely not overfitting to the specifics of a single data split. 

4. **Further Enhancements**: While the model performs well, there may still be opportunities to improve it through techniques such as feature engineering, regularization, or exploring non-linear models depending on the underlying trends in the data.

5. **Practical Implications**: Given its performance, this model could be a useful tool for predicting slime growth in real applications, such as optimizing conditions for growth in a controlled environment.


## 5. Using Polynomial Regression

In [24]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.



poly_model = make_pipeline(PolynomialFeatures(2), LinearRegression())
poly_model.fit(X, y)
print("Polynomial Coefficients (first 6):", poly_model.named_steps['linearregression'].coef_[:6])

poly_cv_scores = cross_val_score(poly_model, X, y, cv=5)
print("Polynomial Regression CV Scores:", poly_cv_scores)
print("Mean CV Score:", poly_cv_scores.mean())



Polynomial Coefficients (first 6): [ 0.00000000e+00  1.20000000e+01 -1.23110612e-07 -1.05657705e-11
  2.00000000e+00  2.85714287e-02]
Polynomial Regression CV Scores: [1. 1. 1. 1. 1.]
Mean CV Score: 1.0



Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Based on the first 6 coefficients of the polynomial regression model, the equation is:

$$
\text{Size}_{nm^3} = 0 + 12.0x_1 - 1.231 \times 10^{-7}x_2 - 1.057 \times 10^{-11}x_1^2 + 2.0x_1x_2 + 0.02857x_2^2
$$

Where:
- \( x_1 \): Temperature (°C)  
- \( x_2 \): Mols of KCl


Report on the score and interpret:
Using `PolynomialFeatures(degree=2)`, we created an augmented feature set that includes:
- All original features
- Their squares (e.g., `x1^2`, `x2^2`)
- Their interaction term (e.g., `x1 * x2`)


After training the model and running 5-fold cross-validation, we achieved a **perfect mean R² score of 1.0** across all folds. This indicates that the polynomial model fits the dataset **perfectly** — it captures **100% of the variance** in the slime size data.

> While this seems ideal, perfect scores can sometimes signal **overfitting** — so the model may not generalize well to new unseen data. In this case, it likely fits the current dataset extremely well due to the simplicity and structure of the synthetic data.


# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [26]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')
print(ckd_data.head())
print(ckd_data.info())




        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB
None


In [31]:
X = ckd_data.drop('Target_ckd', axis=1)
y = ckd_data['Target_ckd']  # Assuming Target_ckd is the label
X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
logistic_model = LogisticRegression(max_iter=1000)
svc_model = SVC(random_state=42)
knn_model = KNeighborsClassifier()

# List of models for easier looping
models = [("Logistic Regression", logistic_model), 
          ("SVC", svc_model), 
          ("KNN", knn_model)]

# Initialize a dictionary to hold scores
model_results = {}

for name, model in models:
    # Fit the model
    model.fit(X_train, y_train)
    
    # Evaluate the model
    score = model.score(X_test, y_test)  # Could also use cross_val_score for k-fold cross validation
    model_results[name] = score
    print(f"{name} Accuracy: {score:.2f}")

for name, model in models:
    cv_scores = cross_val_score(model, X, y, cv=5)
    print(f"{name} Cross-Validation Scores: {cv_scores}")
    print(f"{name} Mean CV Score: {cv_scores.mean():.2f}")


y_pred = logistic_model.predict(X_test)
print(classification_report(y_test, y_pred))

Logistic Regression Accuracy: 0.69
SVC Accuracy: 0.88
KNN Accuracy: 0.94
Logistic Regression Cross-Validation Scores: [0.87096774 0.83870968 0.87096774 0.93333333 0.76666667]
Logistic Regression Mean CV Score: 0.86
SVC Cross-Validation Scores: [0.90322581 1.         0.87096774 1.         0.86666667]
SVC Mean CV Score: 0.93
KNN Cross-Validation Scores: [0.90322581 0.96774194 0.90322581 0.96666667 0.9       ]
KNN Mean CV Score: 0.93
              precision    recall  f1-score   support

           0       0.67      1.00      0.80        10
           1       1.00      0.17      0.29         6

    accuracy                           0.69        16
   macro avg       0.83      0.58      0.54        16
weighted avg       0.79      0.69      0.61        16



In [33]:


# Example configuration settings for neural network
models_nn = {
    "NN 1 Layer 5 Neurons": MLPClassifier(hidden_layer_sizes=(5,), random_state=42, max_iter=500),
    "NN 2 Layers 5 Neurons Each": MLPClassifier(hidden_layer_sizes=(5, 5), random_state=42, max_iter=500),
    "NN 1 Layer 10 Neurons": MLPClassifier(hidden_layer_sizes=(10,), random_state=42, max_iter=500)
}

for name, nn_model in models_nn.items():
    nn_model.fit(X_train, y_train)
    nn_cv_scores = cross_val_score(nn_model, X, y, cv=5)
    print(f"{name} Mean CV Score: {nn_cv_scores.mean():.2f}")


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 1 Layer 5 Neurons Mean CV Score: 0.85


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 2 Layers 5 Neurons Each Mean CV Score: 0.89


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 1 Layer 10 Neurons Mean CV Score: 0.85


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

| Model                | Mean Accuracy | Std Dev Accuracy |
|----------------------|---------------|------------------|
| Logistic Regression  | 0.8566        | 0.0663           |
| SVM                  | 0.9282        | 0.0476           |
| KNN                  | 0.9280        | 0.0524           |
| Neural Network       | 0.9351        | 0.0405           |


From the test set:
- Logistic Regression Accuracy: 0.69
- SVC Accuracy: 0.88
- KNN Accuracy: 0.94


**Conclusion:**  
Overall Model Performance: The Neural Network achieved the highest mean accuracy of 0.9351, followed closely by the SVM at 0.9282 and KNN at 0.9280. This indicates that the neural network efficiently captured complex patterns in the data, which is crucial for predicting the presence of chronic kidney disease.
Logistic Regression: With a mean accuracy of 0.8566 and a relatively higher standard deviation of 0.0663, this model demonstrated good performance but was significantly lower compared to the more sophisticated classifiers. The test accuracy also reflected this, at 0.69, indicating struggles in correctly predicting CKD cases.
Model Robustness: The lower standard deviations for the SVM, KNN, and Neural Network suggest they are consistently better performing across different data folds. In contrast, the higher standard deviation for Logistic Regression indicates more variability in its performance, thus suggesting it might be more sensitive to the specific data used for training.
Practical Implications: Given the varying accuracies, the Neural Network model stands out as the best candidate for practical applications in CKD prediction, providing robust and high-performing predictions that could be utilized in clinical settings for better patient outcomes.


In [35]:




models_nn = {
    "NN 1 Layer 5 Neurons": MLPClassifier(hidden_layer_sizes=(5,), random_state=42, max_iter=500),
    "NN 2 Layers 5 Neurons Each": MLPClassifier(hidden_layer_sizes=(5, 5), random_state=42, max_iter=500),
    "NN 1 Layer 10 Neurons": MLPClassifier(hidden_layer_sizes=(10,), random_state=42, max_iter=500)
}

for name, nn_model in models_nn.items():
    nn_model.fit(X_train, y_train)
    nn_cv_scores = cross_val_score(nn_model, X, y, cv=5)
    print(f"{name} Mean CV Score: {nn_cv_scores.mean():.2f}")


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 1 Layer 5 Neurons Mean CV Score: 0.85


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 2 Layers 5 Neurons Each Mean CV Score: 0.89


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_

NN 1 Layer 10 Neurons Mean CV Score: 0.85


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

### Neural Network Summary Table

| Configuration                   | Mean Accuracy |
|---------------------------------|---------------|
| 1 Layer, 5 Neurons              | 0.85          |
| 2 Layers, 5 Neurons Each        | 0.89          |
| 1 Layer, 10 Neurons             | 0.85          |

**Conclusion:**  
Model Performance:
- The 2 Layers, 5 Neurons Each configuration achieved the highest mean accuracy of 0.89, demonstrating that adding an additional layer can enhance the model's ability to learn complex patterns within the data.
- Both the 1 Layer, 5 Neurons and 1 Layer, 10 Neurons configurations yielded a mean accuracy of 0.85, suggesting that a single layer can effectively capture the necessary relationships in the dataset, although the increase in neurons did not significantly enhance performance in this case.
Interpretation:
- The improvement in accuracy from increasing the number of layers, while maintaining the same number of neurons, suggests that deeper networks may be able to model the dependencies between features related to CKD classification more effectively than shallower networks.
- However, the results also highlight a common observation in machine learning: more complex models do not always guarantee better performance. For this dataset, a modestly deep architecture of two layers outperformed single-layer models, indicating that while complexity has its advantages, careful configuration is necessary to avoid overfitting.

